src/main/resources/com/tresys/magvar/xsd/magvar.dfdl.xsd

```diff
@@ -50,28 +50,37 @@ SOFTWARE.
         emptyValueDelimiterPolicy="both" representation="text" textNumberRep="standard"
         lengthKind="implicit" encoding="ASCII" encodingErrorPolicy="error" outputNewLine="%CR;%LF;" 
         textNumberJustification="right" textNumberPadCharacter="%SP;" textTrimKind="padChar" textPadKind="padChar"
-        textNumberPattern="###0.0##;-###0.0##" textStandardZeroRep="" />
+        textStandardZeroRep="" />
+      <dfdl:defineFormat name="floatFmt">
+        <dfdl:format textNumberPattern="###0.0##;-###0.0##"/>
+      </dfdl:defineFormat>
     </xs:appinfo>
   </xs:annotation>
 
   <xs:element name="magvar" dfdl:terminator="999999999999999999999999999999999999999999999999%NL;999999999999999999999999999999999999999999999999%NL;">
     <xs:complexType>
       <xs:sequence dfdl:separator="%NL;" dfdl:separatorPosition="postfix">
-        <xs:element name="header" type="xs:string" dfdl:lengthKind="delimited"/>
+        <xs:element name="header" type="tns:paddedString" dfdl:lengthKind="delimited"/>
         <xs:element name="record" maxOccurs="unbounded">
           <xs:complexType>
             <xs:sequence dfdl:separator="%SP;">
               <xs:element name="item1" type="xs:int" dfdl:lengthKind="explicit" dfdl:length="3"/>
               <xs:element name="item2" type="xs:int" dfdl:lengthKind="explicit" dfdl:length="2"/>
-              <xs:element name="item3" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="9"/>
-              <xs:element name="item4" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="9"/>
-              <xs:element name="item5" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="10"/>
-              <xs:element name="item6" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="10"/>
+              <xs:element name="item3" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="9" dfdl:ref="tns:floatFmt"/>
+              <xs:element name="item4" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="9" dfdl:ref="tns:floatFmt"/>
+              <xs:element name="item5" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="10" dfdl:ref="tns:floatFmt"/>
+              <xs:element name="item6" type="xs:double" dfdl:lengthKind="explicit" dfdl:length="10" dfdl:ref="tns:floatFmt"/>
             </xs:sequence>
           </xs:complexType>
         </xs:element> 
       </xs:sequence>
     </xs:complexType>
   </xs:element>
+  

```

Comment:
I had to introduce this simple type in order to express the minLength facet. IBM DFDL insists on this facet because the header element above has lengthKind 'delimited' and textPadKind='padChar'. So it needs a minimum length to pad to.  However, the XSD minLength facet is not a DFDL property (which don't have default values generally), it is an XSD thing, which defaults to zero. So I think insisting on this facet is not actually a DFDL requirement.

Reply:
@mbeckerle If `minLength` is 0, then no padding will ever take place, so you could set `textPadKind="none"` instead.

I don't think schema facets in general have default values. For example what would you default `length` to?  Our code uses `org.eclipse.xsd` and makes a call to `getEffectiveMinLengthFacet()` on the element's simple type (anonymous or explicit) which returns `null` if there is no `minLength` facet anywhere in the type hierarchy. And that `null` is what causes the validation error you saw. OK so we could assume `minLength=0` and `maxLength=Inf` but the point here is that the elements say (or inherits) `textPadKind="padChar" so the user could have omitted the `minLength` by mistake, so at least a warning needs emitting. 

Note that we solve this in our big standard schemas by creating a set of explicit simple types and never using anonymous simple types for elements.

Reply:
I cannot find anything explicit about minLength defaulting to zero. There is the statement about restrictions that any new minLength must be greater than or equal to the minLength of the base. That can be extrapolated to the minLength being zero to start with. A warning is sensible here as you point out. Whether this should be SDE or not is debatable. This schema is correct either way though as it is now.